### DISCUSSION TITLE
为什么 MySQL 使用 B+ 树· Why's THE Design? · /whys-the-design-mysql-b-plus-tree

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#579] [Announcements]
https://draveness.me/whys-the-design-mysql-b-plus-tree

**Comment by hhyvs111** [comment]
一个面试题，想问一下博主，一条普通的sql语句，如
```
t1 
id
a b c d

select * from t1 where c = 5;
```
id是主键，然后有c的索引，这一条语句会访问几次磁盘？

我的理解是b+树的高度就是访问磁盘的次数，也就是说先用3次（假如树的高度为3）找到c索引的键值，然后再从主键索引里找出对应的数据，一共就是6次。这样理解对吗？

**Comment by draveness** [comment]
> 一个面试题，想问一下博主，一条普通的sql语句，如
> 
> ```
> t1 
> id
> a b c d
> 
> select * from t1 where c = 5;
> ```
> 
> id是主键，然后有c的索引，这一条语句会访问几次磁盘？
> 
> 我的理解是b+树的高度就是访问磁盘的次数，也就是说先用3次（假如树的高度为3）找到c索引的键值，然后再从主键索引里找出对应的数据，一共就是6次。这样理解对吗？

你可以这么理解，但是 MySQL 中一般情况下所有的非叶节点都会在内存中，只有访问叶节点才会需要随机 I/O 加载数据，如果不考虑这个就是 6 次，但是一般情况下最多只会有 2 次

**Comment by draveness** [comment]
> MySQL中Page存在于Buffer Pool和是否为叶子节点没有关系，按照你的描述，这里只能假设Page均不在Buffer Pool中，所以需要6次IO: 二级索引3次，主键索引3次.

对于大表所有的索引页都会缓存在内存里，但是内存里一般是放不下整个表中全部数据的。

>  在InnoDB中的实现中Buffer Poool会进行read ahead预读，所以真正具体IO次数无法确定.

你说得对，不过文中说过了不考虑优化策略。

**Comment by draveness** [comment]
> 这个我看代码倒没有发现这方面的优化，本质上Buffer Pool是LRU策略，针对大表缓存non leaf page和LRU会冲突，并且Buffer Pool对表的大小也是无感知的.

与整个表的大小相比，索引页的大小完全可以存储在内存中，而访问索引页的频率会大概率比数据页高得多，所以只要内存足够，最终索引页一定会缓存在内存中。

**Comment by leviathan0992** [comment]
> @draveness
> > 这个我看代码倒没有发现这方面的优化，本质上Buffer Pool是LRU策略，针对大表缓存non leaf page和LRU会冲突，并且Buffer Pool对表的大小也是无感知的.
> 
> 与整个表的大小相比，索引页的大小完全可以存储在内存中，而访问索引页的频率会大概率比数据页高得多，所以只要内存足够，最终索引页一定会缓存在内存中。

你说的是可能的结果，MySQL没有这样的机制.

**Comment by draveness** [comment]
> @Leviathan1995
> > @draveness
> > > 这个我看代码倒没有发现这方面的优化，本质上Buffer Pool是LRU策略，针对大表缓存non leaf page和LRU会冲突，并且Buffer Pool对表的大小也是无感知的.
> > 
> > 与整个表的大小相比，索引页的大小完全可以存储在内存中，而访问索引页的频率会大概率比数据页高得多，所以只要内存足够，最终索引页一定会缓存在内存中。
> 
> 你说的是可能的结果，MySQL没有这样的机制.

行 :+1:

**Comment by arrayadd** [comment]
还是挺感慨的，软件工程师们想尽各种办法来做优化，性能上才能提升那么一点点，而硬件上的一旦有进步，比起来简直就是小巫见大巫。顺序读取1MB数据，机械硬盘是20ms，SSD只需要1ms.  20倍差距，B,B+优劣在这里就.....

**Comment by MikeoPerfect** [comment]
基本上讲到点子上了，MySQL的InnoDB存储引擎，之所以采用B+树的设计思路，和InnoDB的数据存储方式离不开关系，操作系统的页默认大小为4KB，innoDB的页大小为16KB，当表中的记录存入InnoDB时，为了查询方便会采用操作系统文件管理中的文件索引思路存放记录数据，也即文中介绍的指针型节点，一个记录从访问到被访问基本会经过如下过程：目录页、页目录、页、记录，其中前三者体现的就是索引的价值，最后一者就是我们说的数据了，而这就是B+树为什么非叶子节点只有索引，叶子节点才有数据的根源

**Comment by MikeoPerfect** [comment]
> 今天碰巧看到一篇博文，介绍的就是MySQL innodb_page_size，可能对大家有所帮助，
innodb_page_size是一个初始化数据库实例的参数，在目前的版本中（>=5.7.6），可以选择的值有4096, 8192, 16384, 32768, 65536。默认是16KB

一般越小，内存划分粒度越大，使用率越高，但是会有其他问题，就是限制了索引字段还有整行的大小。innodb引擎读取内存还有更新都是一页一页更新的，这个innodb_page_size决定了，一个基本页的大小。常用B+Tree索引，B+树是为磁盘及其他存储辅助设备而设计一种平衡查找树（不是二叉树）。B+树中，所有记录的节点按大小顺序存放在同一层的叶子节点中，各叶子节点用指针进行连接。MySQL将每个叶子节点的大小设置为一个页的整数倍，利用磁盘的预读机制，能有效减少磁盘I/O次数，提高查询效率。 如果一个行数据，超过了一页的一半，那么一个页只能容纳一条记录，这样B+Tree在不理想的情况下就变成了双向链表。 

 原文出处：[MySQL原理 - InnoDB表的限制](https://zhuanlan.zhihu.com/p/79987871)

**Comment by draveness** [comment]
> 今天碰巧看到一篇博文，介绍的就是MySQL innodb_page_size，可能对大家有所帮助，
>
> > innodb_page_size是一个初始化数据库实例的参数，在目前的版本中（>=5.7.6），可以选择的值有4096, 8192, 16384, 32768, 65536。默认是16KB
> > 一般越小，内存划分粒度越大，使用率越高，但是会有其他问题，就是限制了索引字段还有整行的大小。innodb引擎读取内存还有更新都是一页一页更新的，这个innodb_page_size决定了，一个基本页的大小。常用B+Tree索引，B+树是为磁盘及其他存储辅助设备而设计一种平衡查找树（不是二叉树）。B+树中，所有记录的节点按大小顺序存放在同一层的叶子节点中，各叶子节点用指针进行连接。MySQL将每个叶子节点的大小设置为一个页的整数倍，利用磁盘的预读机制，能有效减少磁盘I/O次数，提高查询效率。 如果一个行数据，超过了一页的一半，那么一个页只能容纳一条记录，这样B+Tree在不理想的情况下就变成了双向链表。
>
> > 原文出处：[MySQL原理 - InnoDB表的限制](https://zhuanlan.zhihu.com/p/79987871)

之前写过 MySQL 和 InnoDB 的文章，可以看下 [『浅入浅出』MySQL 和 InnoDB](https://draveness.me/mysql-innodb)

**Comment by XhinLiang** [comment]
博主，你的博客貌似挂了 https://draveness.me

**Comment by draveness** [comment]
> 博主，你的博客貌似挂了 https://draveness.me

你是北京联通么，北京联通把我的博客劫持了，我跟他们投诉了，你可以先访问 https://draven.co

**Comment by boykait** [comment]
所以在遇到可能存在相同键 author = 'draven 这个hash应该都会散列到同一个hash桶里面吧

**Comment by XhinLiang** [comment]
> @boykait
> 所以在遇到可能存在相同键 author = 'draven 这个hash应该都会散列到同一个hash桶里面吧

author 是主键的话，是的。

**Comment by AngusLean** [comment]
后面写到:
> 从今天的角度来看，B+ 树可能不是 InnoDB 的最优选择

这个作何解释，与SSD这种新的存储方式有关还是？

**Comment by draveness** [comment]
> @AngusLean
> 后面写到:
> > 从今天的角度来看，B+ 树可能不是 InnoDB 的最优选择
> 
> 这个作何解释，与SSD这种新的存储方式有关还是？

这里不是说 B+ 树一定有问题，只是说其他的数据结构也可以作为关系型数据库的底层数据结构，例如 LSM 树。

**Comment by galaxy001** [comment]
从今天的角度来看，B+ 树可能不是 InnoDB 的最优选择

好奇今天来看，有哪些更好的选择？

**Comment by draveness** [comment]
> ```
> 从今天的角度来看，B+ 树可能不是 InnoDB 的最优选择
> ```
> 
> 好奇今天来看，有哪些更好的选择？

你看看上面？

**Comment by galaxy001** [comment]
> @draveness
> > ```
> > 从今天的角度来看，B+ 树可能不是 InnoDB 的最优选择
> > ```
> > 
> > 好奇今天来看，有哪些更好的选择？
> 
> 你看看上面？

上面提到的几种都说了不如B+树呀？B tree不行，hash不行，三个一起上也说了有缺点。

**Comment by draveness** [comment]
> > @draveness
> > > ```
> > > 从今天的角度来看，B+ 树可能不是 InnoDB 的最优选择
> > > ```
> > > 
> > > 
> > > 好奇今天来看，有哪些更好的选择？
> > 
> > 
> > 你看看上面？
> 
> 上面提到的几种都说了不如B+树呀？B tree不行，hash不行，三个一起上也说了有缺点。

我说的是评论，LSM 树

> 这里不是说 B+ 树一定有问题，只是说其他的数据结构也可以作为关系型数据库的底层数据结构，例如 LSM 树。

**Comment by galaxy001** [comment]
LSM 树是一种专门用来优化写入的数据结构，它将随机写变成了顺序写显著地提高了写入性能，但是却牺牲了读的效率，这与大多数场景需要的特点是不匹配的，所以 MongoDB 最终还是选择读取性能更好的 B 树作为默认的数据结构；

看到了，然后跑去翻了下隔壁的说明。

就是说，考虑到读取，LSM的写入优势对InnoDB来说收益不大。
所以InnoDB还是不能选LSM，得维持B+。

咱就说开源软件如果有必要，完全可以添加支撑新的组件嘛。
就像btrfs遇到比lz4更好的zstd，就添加zstd压缩支持了。并没有因为历史原因，不能使用其他格式。

那么，现在做需要遍历的key-value存储，还是只能用B+树了。

**Comment by iapolog2u** [comment]
我对比另外一篇文章的B+树介绍，和您所画的B+ Tree好像不一样，您画的B+树在下面文章里好像是B树

文章地址：https://zhuanlan.zhihu.com/p/27700617

**Comment by draveness** [comment]
> @ilp2020
> 我对比另外一篇文章的B+树介绍，和您所画的B+ Tree好像不一样，您画的B+树在下面文章里好像是B树
> 
> 文章地址：https://zhuanlan.zhihu.com/p/27700617

具体是哪里不一样呢？

**Comment by neal-zhu** [comment]
请问最后一张图，B+ tree 的 leaf node 里面，为什么还有类似3，5，7，8 这样的 key ？

**Comment by draveness** [comment]
> 请问最后一张图，B+ tree 的 leaf node 里面，为什么还有类似3，5，7，8 这样的 key ？

为什么不应该有呢，在我有限的认知里，我认为这样是没有问题的。如果你觉得这里确实存在错误，可以给一些理由和 Reference，要不然我也不知道从哪里开始回答你的问题。

**Comment by neal-zhu** [comment]
leaf node 中应该都是每个具体的 row 了。最后一张图的 leaf node 为什么看着有点像是 internal node 和 leaf node 的结合？3，5，7， 这种 row 对应的数据在哪里呢？

**Comment by draveness** [comment]
> leaf node 中应该都是每个具体的 row 了。最后一张图的 leaf node 为什么看着有点像是 internal node 和 leaf node 的结合？3，5，7， 这种 row 对应的数据在哪里呢？

下面那个黑红的才是叶节点...

**Comment by neal-zhu** [comment]
> @draveness
> > leaf node 中应该都是每个具体的 row 了。最后一张图的 leaf node 为什么看着有点像是 internal node 和 leaf node 的结合？3，5，7， 这种 row 对应的数据在哪里呢？
> 
> 下面那个黑红的才是叶节点...

哦 我重新看了一遍反应过来了

**Comment by lvseouren** [comment]
棒

**Comment by Jesse-so-cool** [comment]
博主你好 有点疑问：
问题1：
 我看到网上别的文章对b树的说法是，b树非叶子节点也会存数据，所以会导致树高度增加，从而导 
 致的额外随机i/o。
 但博主的文章是， B+ 树这种数据结构会增加树的高度

问题2：

 另外，文章说到，B+ 树的所有叶节点可以通过指针相互连接，能够减少顺序遍历时产生的额外随机 I/O
 这么说，B+和B树的效率的最大影响是B+树可以通过『指针』依次按顺序连接，这样理解有问题么

**Comment by draveness** [comment]
> 博主你好 有点疑问：
> 问题1：
> 我看到网上别的文章对b树的说法是，b树非叶子节点也会存数据，所以会导致树高度增加，从而导
> 致的额外随机i/o。
> 但博主的文章是， B+ 树这种数据结构会增加树的高度

这个能给出来链接么，质量高一些的引用

> 问题2：
> 
> 另外，文章说到，B+ 树的所有叶节点可以通过指针相互连接，能够减少顺序遍历时产生的额外随机 I/O
> 这么说，B+和B树的效率的最大影响是B+树可以通过『指针』依次按顺序连接，这样理解有问题么

是的，连接比较方便一些

**Comment by Jesse-so-cool** [comment]
随便找了一个知乎专栏的 https://zhuanlan.zhihu.com/p/81273236

文章回答：
因为B树不管叶子节点还是非叶子节点，都会保存数据，这样导致在非叶子节点中能保存的指针数量变少（有些资料也称为扇出）

指针少的情况下要保存大量数据，只能增加树的高度，导致IO操作变多，查询性能变低；

**Comment by Jesse-so-cool** [comment]
> > 博主你好 有点疑问：
> > 问题1：
> > 我看到网上别的文章对b树的说法是，b树非叶子节点也会存数据，所以会导致树高度增加，从而导
> > 致的额外随机i/o。
> > 但博主的文章是， B+ 树这种数据结构会增加树的高度
> 
> 这个能给出来链接么，质量高一些的引用
> 
> > 问题2：
> > 另外，文章说到，B+ 树的所有叶节点可以通过指针相互连接，能够减少顺序遍历时产生的额外随机 I/O
> > 这么说，B+和B树的效率的最大影响是B+树可以通过『指针』依次按顺序连接，这样理解有问题么
> 
> 是的，连接比较方便一些

随便找了一个知乎专栏的 https://zhuanlan.zhihu.com/p/81273236

文章回答：
因为B树不管叶子节点还是非叶子节点，都会保存数据，这样导致在非叶子节点中能保存的指针数量变少（有些资料也称为扇出）

指针少的情况下要保存大量数据，只能增加树的高度，导致IO操作变多，查询性能变低；

求解答

**Comment by xwind** [comment]
感觉对 b 和 b+ 的分析优势说得还不够透彻。文章中标注的“通过『指针』依次按顺序连接”其实有一点误导性，b 树也可以用 O(nlogn) 的代价使用指针把对应的 next node 串联起来，但是这种实现在旋转时不能保证 b 树的树节点保持在一段连续的区块内。因此，b+ 树更本质得目的在于使用额外 O(n) 的指针索引空间，换取了数据在内存或者硬盘的物理数据连续性，保证了数据连续性，就有了后面分析的两个性质。

(您这 blog 文章质量和数量这么高，看来架构组的工作 work and life balance 做得很好呀，哈哈哈，逃……

**Comment by Jesse-so-cool** [comment]
> > @Jesse-so-cool
> > > > 博主你好 有点疑问：
> > > > 问题1：
> > > > 我看到网上别的文章对b树的说法是，b树非叶子节点也会存数据，所以会导致树高度增加，从而导
> > > > 致的额外随机i/o。
> > > > 但博主的文章是， B+ 树这种数据结构会增加树的高度
> > > 
> > > 
> > > 这个能给出来链接么，质量高一些的引用
> > > > 问题2：
> > > > 另外，文章说到，B+ 树的所有叶节点可以通过指针相互连接，能够减少顺序遍历时产生的额外随机 I/O
> > > > 这么说，B+和B树的效率的最大影响是B+树可以通过『指针』依次按顺序连接，这样理解有问题么
> > > 
> > > 
> > > 是的，连接比较方便一些
> > 
> > 
> > 随便找了一个知乎专栏的 https://zhuanlan.zhihu.com/p/81273236
> > 文章回答：
> > 因为B树不管叶子节点还是非叶子节点，都会保存数据，这样导致在非叶子节点中能保存的指针数量变少（有些资料也称为扇出）
> > 指针少的情况下要保存大量数据，只能增加树的高度，导致IO操作变多，查询性能变低；
> > 求解答
> 
> 对 我也看到过 我认为也是的~ 因为每一个页的存储空间是有限的 如果存的了数据那必然存储的key就会变少 这样必然会导致 树的层级别 变高 IO的读取边多

其实我想问作者的是关于b树查询性能较低的原因，
有些博文是说的 指针少的情况下要保存大量数据，只能增加树的高度，导致IO操作变多，查询性能变低
作者说的是 由于指针问题，且b树会保存节点数据，所以查询的时候会导致随机I/O增加，导致性能降低
说法有些不一致

**Comment by handsomemantony** [comment]
有个疑问,为什么数据是选择b+树而部署skiplist来存储呢?跳表能做到无锁,查询性能与b+树差不多,岂不是更好?

**Comment by draveness** [comment]
> 有个疑问,为什么数据是选择b+树而部署skiplist来存储呢?跳表能做到无锁,查询性能与b+树差不多,岂不是更好?

B+ 树更适合磁盘访问和存储

**Comment by techone577** [comment]
您好，《为什么 MongoDB 使用 B 树》这篇文章是已经删除了吗

**Comment by draveness** [comment]
> 您好，《为什么 MongoDB 使用 B 树》这篇文章是已经删除了吗

这篇文章有问题，没时间改暂时关掉了

**Reply by licolas94** [reply]
3年过去了～～，刚想问为什么那篇文章打不开。。。

**Comment by techone577** [comment]
> @draveness
> > 您好，《为什么 MongoDB 使用 B 树》这篇文章是已经删除了吗
> 
> 这篇文章有问题，没时间改暂时关掉了

好的感谢解答，我看现在网上包括官方文档对 MongoDB 使用 B 树还是 B+ 树的描述都有些不清晰，所以期待大佬修改后的文章

**Comment by xiaoheiAh** [comment]
`{% include related/whys-the-design.md %}` 关联的子页面失效了

**Comment by ceezyyy** [comment]
大佬，B+ tree 为啥叶子结点没有 6 ？

**Comment by draveness** [comment]
> @ceezyyy 大佬，B+ tree 为啥叶子结点没有 6 ？

因为没人写所以没有？

**Comment by rookiezq** [comment]
想问一下，既然层与层之间会有I/O，那在遍历叶子节点时会有I/O吗？叶子节点时连续存储的吗？谢谢！

**Comment by nobject** [comment]
一直没搞明白树的层数与磁盘io的关系，比如红黑树层数高，磁盘的io就高，是因为同一层的节点容易存在同一个数据页中么？

**Comment by rwxe** [comment]
> @nobject
> 一直没搞明白树的层数与磁盘io的关系，比如红黑树层数高，磁盘的io就高，是因为同一层的节点容易存在同一个数据页中么？

我和你有同样的问题，目前也没搞明白，“为什么一个硬盘块只能存储一个树节点”。希望有博主或其他懂的大佬能解答我的疑惑。

**Comment by draveness** [comment]
> > @nobject
> > 一直没搞明白树的层数与磁盘io的关系，比如红黑树层数高，磁盘的io就高，是因为同一层的节点容易存在同一个数据页中么？
> 
> 我和你有同样的问题，目前也没搞明白，“为什么一个硬盘块只能存储一个树节点”。希望有博主或其他懂的大佬能解答我的疑惑。

你希望访问一个节点的时候触发多次 IO 么

**Comment by rookiezq** [comment]
> @nobject
> 一直没搞明白树的层数与磁盘io的关系，比如红黑树层数高，磁盘的io就高，是因为同一层的节点容易存在同一个数据页中么？

不是的，磁盘页大小是4k，mysql的页是16k，所以mysql一页≈4页磁盘页；mysql中按照B+树存储，每一个树节点占了一个mysql页，所以当一个节点指向另一个节点时就会发生io

**Comment by rwxe** [comment]
> > > @nobject
> > > 一直没搞明白树的层数与磁盘io的关系，比如红黑树层数高，磁盘的io就高，是因为同一层的节点容易存在同一个数据页中么？
> > 
> > 
> > 我和你有同样的问题，目前也没搞明白，“为什么一个硬盘块只能存储一个树节点”。希望有博主或其他懂的大佬能解答我的疑惑。
> 
> 你希望访问一个节点的时候触发多次 IO 么

不是很明白您的意思。我的意思是，为什么不能用二叉树，将多个节点储存到一个块中，每个节点的指针可以设计为`块号+偏移`。每次访问一个节点，和它同属一个磁盘块的节点也能一同读出。为什么一个磁盘块只能存储至多一个节点，这个我弄不懂。

**Reply by Yuu177** [reply]
我的问题和你一样，但是回复到这里就断了，这个疑惑你解决了吗？

**Comment by zhouwentong1993** [comment]
> @draveness
> > 博主你好 有点疑问：
> > 问题1：
> > 我看到网上别的文章对b树的说法是，b树非叶子节点也会存数据，所以会导致树高度增加，从而导
> > 致的额外随机i/o。
> > 但博主的文章是， B+ 树这种数据结构会增加树的高度
> 
> 这个能给出来链接么，质量高一些的引用
> 
> > 问题2：
> > 
> > 另外，文章说到，B+ 树的所有叶节点可以通过指针相互连接，能够减少顺序遍历时产生的额外随机 I/O
> > 这么说，B+和B树的效率的最大影响是B+树可以通过『指针』依次按顺序连接，这样理解有问题么
> 
> 是的，连接比较方便一些

博主好，《数据密集型应用系统设计》中的数据库核心：数据结构一章中，在「优化 B-tree」章节中提到了 B+ 树的层数少于 B 树。见 P83。

**Comment by zhangyuanxue** [comment]
前文降到讲到计```算机在读写文件时会以页为单位将数据加载到内存中```，后文提到一种B树的 4 次磁盘的随机 I/O 
```
加载根节点所在的页，发现根节点的第一个元素是 6，大于 4；
通过根节点的指针加载左子节点所在的页，遍历页面中的数据，找到 5；
重新加载根节点所在的页，发现根节点不包含第二个元素；
通过根节点的指针加载右子节点所在的页，遍历页面中的数据，找到 7 和 8；
```
从上面这段话可以看出B数的每一层是在不同的页里的，这一点大佬并没有提及，可否补充点细节呢？不太明白为什么不同层在不同的页

**Comment by zhangyuanxue** [comment]
您发的邮件我已收到，我会马上查看的 ，谢谢。

**Comment by zhangyuanxue** [comment]
您发的邮件我已收到，我会马上查看的 ，谢谢。

**Comment by Juminiy** [comment]
MacBook m1一页是16KB

**Comment by songzhibin97** [comment]
`而 B 树和 B 树和 B+` 这里重复了

**Comment by zhangyuanxue** [comment]
您发的邮件我已收到，我会马上查看的 ，谢谢。

**Comment by gama-ray** [comment]
相对于B树，B+树一个更显著的优势，由于非叶子节点不存储数据，一个节点内就有更多的空间做更细粒度的区间划分，这对于确定目标数据所在位置的效率是非常有益的

**Comment by zhangyuanxue** [comment]
您发的邮件我已收到，我会马上查看的 ，谢谢。
